### DISCUSSION TITLE
References to check out

### DISCUSSION THREAD

**Discussion by DavidPratten** [hydromatic/morel#98] [General]
Tannen, V., Buneman, P., & Naqvi, S. (1992). [Structural recursion as a query language](https://www.pure.ed.ac.uk/ws/portalfiles/portal/16517467/Structural_Recursion_as_a_Query_Language.pdf). Technical Reports (CIS), 403. 
> Abstract
We propose a programming paradigm that tries to get close to both the semantic simplicity of relational
algebra, and the expressive power of unrestricted programming languages. Its main computational
engine is structural recursion on sets. ... The uniform paradigm of the language suggests positive expectations for the
optimization problem. In fact, structural recursion yields finer grain programming. Therefore we expect
that lower-level and therefore better optimizations will be feasible.

**Comment by julianhyde** [comment]
[SQL, Homomorphisms and Constraint Satisfaction Problems](https://www.philipzucker.com/sql_graph_csp/), @philzook58 (2024)
